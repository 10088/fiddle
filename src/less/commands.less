@import "variables.less";

.commands {
  display: flex;
  justify-content: space-between;
  height: 40px;
  padding-top: 12px;
  font-family: var(--default-fonts);

  button, input, select {
    -webkit-app-region: no-drag;
  }

  select {
    width: 100px;
    font-size: 12px;
    height: 28px;
    background-color: white;
    margin-left: 5px;
    line-height: 15px;
  }

  &>div {
    display: flex;

    &:first-of-type {
      button {
        margin-left: 80px;
      }
    }

    &::last-of-type {
      margin-right: 10px;

      button {
        margin-left: 5px;
      }
    }
  }

  .address-bar {
    height: 28px;
    width: 360px;
    transition: width .4s;
    display: flex;

    &.empty {
      width: 200px;
    }

    &::focus-within {
      span, input {
        border-color: @electron-bright;
      }
    }

    input {
      flex-grow: 1;
      background: white;
      border-left: 0;
      border-top-left-radius: 0;
      border-bottom-left-radius: 0;
      padding-left: 0;
      height: 28px;

      &:focus {
        border-top-left-radius: 0;
        border-bottom-left-radius: 0;
        border-left: 0;
      }
    }

    span {
      background: white;
      color: #1d1d1d;
      border-bottom-color: rgb(209, 209, 209);
      border-bottom-style: solid;
      border-bottom-left-radius: 4px;
      border-bottom-width: 1px;
      border-image-outset: 0px;
      border-image-repeat: stretch;
      border-image-slice: 100%;
      border-image-source: none;
      border-image-width: 1;
      border-left-color: rgb(209, 209, 209);
      border-left-style: solid;
      border-left-width: 1px;
      border-top-color: rgb(209, 209, 209);
      border-top-left-radius: 4px;
      border-top-style: solid;
      border-top-width: 1px;
      color: rgb(0, 0, 0);
      cursor: auto;
      display: inline-block;
      font-family: system-ui;
      font-size: 11px;
      height: 28px;
      letter-spacing: normal;
      line-height: 16px;
      margin: 0 6px 0 0;
      padding: 6px 6px 10px 10px;
      user-select: text;
      width: 132px;
      -webkit-app-region: no-drag;
      text-rendering: auto;
      text-shadow: none;
      text-transform: none;
    }
  }
}
