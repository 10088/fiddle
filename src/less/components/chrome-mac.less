.darwin {
  .chrome {
    text-align: center;
    background-color: @background-1;
    border-bottom: @border;
    -webkit-app-region: drag;
    height: 23px;
    font-size: 1.4rem;

    small,
    p,
    span {
      user-select: none;
    }
  }
}
