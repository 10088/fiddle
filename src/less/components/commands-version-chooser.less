button.version-chooser.bp3-button {
  border-radius: 3px;
}
