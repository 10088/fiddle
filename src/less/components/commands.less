@import "../variables.less";

header {
  box-shadow: 0 0 0 1px rgba(16, 22, 26, 0.2), 0 0 0 rgba(16, 22, 26, 0), 0 1px 1px rgba(16, 22, 26, 0.4);
  background-color: @background-1;
  -webkit-app-region: drag;
}

.commands {
  display: flex;
  justify-content: space-between;
  height: 50px;
  padding: 10px;
  box-sizing: border-box;
  font-family: @fonts-common;

  button,
  input,
  select {
    -webkit-app-region: no-drag;
    margin-bottom: 0;
  }

  select {
    width: 100px;
    margin-left: 5px;
  }

  & > div {
    display: flex;

    &:last-of-type {
      margin-right: 10px;

      button {
        margin-left: 5px;
      }
    }
  }

  .address-bar {
    height: 28px;
    width: 360px;
    transition: width 0.4s;
    display: flex;

    input,
    span {
      font-family: @fonts-common;
      font-weight: 300;
      font-size: 11px;
      cursor: auto;
      -webkit-app-region: no-drag;
    }

    input {
      flex-grow: 1;
      background: white;
      padding-left: 1rem;
      height: 2.6rem;
    }

    input:invalid {
      box-shadow: inset 0 0 0 1px #e4252d;
    }

    span {
      background: white;
      color: #1d1d1d;
      border-bottom-color: rgb(209, 209, 209);
      border-bottom-style: solid;
      border-bottom-left-radius: 4px;
      border-bottom-width: 1px;
      border-image-outset: 0;
      border-image-repeat: stretch;
      border-image-slice: 100%;
      border-image-source: none;
      border-image-width: 1;
      border-left-color: rgb(209, 209, 209);
      border-left-style: solid;
      border-left-width: 1px;
      border-top-color: rgb(209, 209, 209);
      border-top-left-radius: 4px;
      border-top-style: solid;
      border-top-width: 1px;
      display: inline-block;
      height: 1.2rem;
      letter-spacing: normal;
      margin: 0;
      padding: 6px 0 10px 10px;
      user-select: text;
      width: 132px;
      text-rendering: auto;
      text-shadow: none;
      text-transform: none;
    }

    &.empty {
      width: 200px;
    }

    &:focus-within {
      span,
      input {
        border-color: @foreground-2;
      }
    }
  }
}
