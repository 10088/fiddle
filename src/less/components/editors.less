.mosaic-tile {
  animation: fadein 0.15s;
}

.mosaic-window-toolbar  {
  & > div {
    display: flex;
    justify-content: space-between;
    width: 100%;
    padding: 0 3px 0 29px;

    .mosaic-controls {
      transform: scale(0.7);
    }

    h5 {
      margin: 4px 0 0 0;
    }

    button {
      margin-left: 8px;
    }
  }
}

@keyframes fadein {
  from { opacity: 0; }
  to { opacity: 1; }
}
