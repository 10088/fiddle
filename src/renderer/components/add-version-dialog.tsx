import { observer } from 'mobx-react';
import * as path from 'path';
import * as React from 'react';
import * as semver from 'semver';

import { NpmVersion } from '../../interfaces';
import { getElectronNameForPlatform } from '../../utils/electron-name';
import { AppState } from '../state';
import { Dialog } from './dialog';

export interface AddVersionDialogProps {
  appState: AppState;
}

export interface AddVersionDialogState {
  isValidElectron: boolean;
  isValidVersion: boolean;
  file?: File;
  version: string;
}

/**
 * The "add version" dialog allows users to add custom builds of Electron.
 *
 * @class AddVersionDialog
 * @extends {React.Component<AddVersionDialogProps, AddVersionDialogState>}
 */
@observer
export class AddVersionDialog extends React.Component<AddVersionDialogProps, AddVersionDialogState> {
  constructor(props: AddVersionDialogProps) {
    super(props);

    this.state = {
      isValidVersion: false,
      isValidElectron: false,
      version: ''
    };

    this.onSubmit = this.onSubmit.bind(this);
    this.onClose = this.onClose.bind(this);
    this.onChangeFile = this.onChangeFile.bind(this);
    this.onChangeVersion = this.onChangeVersion.bind(this);
  }

  /**
   * Handles a change of the file input
   *
   * @param {React.ChangeEvent<HTMLInputElement>} event
   */
  public async onChangeFile(event: React.ChangeEvent<HTMLInputElement>) {
    const { files } = event.target;
    const { binaryManager } = this.props.appState;
    const file = files && files[0] ? files[0] : undefined;

    const isValidElectron = !!(file
      && file.path
      && await binaryManager.getIsDownloaded('custom', file.path));

    this.setState({ file, isValidElectron });
  }

  public onChangeVersion(event: React.ChangeEvent<HTMLInputElement>) {
    const version = event.target.value || '';
    const isValidVersion = !!semver.valid(version);

    this.setState({
      version,
      isValidVersion,
     });
  }

  /**
   * Handles the submission of the dialog
   *
   * @returns {Promise<void>}
   */
  public async onSubmit(): Promise<void> {
    const { file, version } = this.state;

    if (!file) return;

    const name = file.path
      .slice(-20)
      .split(path.sep)
      .slice(1)
      .join(path.sep);

    const toAdd: NpmVersion = {
      localPath: file.path,
      version,
      name
    };

    this.props.appState.addLocalVersion(toAdd);
    this.onClose();
  }

  /**
   * Closes the dialog
   */
  public onClose() {
    this.props.appState.isAddVersionDialogShowing = false;
    this.reset();
  }

  /**
   * Reset this component's state
   */
  public reset(): void {
    this.setState({
      isValidElectron: false,
      isValidVersion: false,
      version: '',
      file: undefined
    });
  }

  get buttons() {
    const canSubmit = this.state.isValidElectron && this.state.isValidVersion;

    return [
      (
        <button
          className='button'
          key='submit'
          disabled={!canSubmit}
          onClick={this.onSubmit}
        >
          Add
        </button>
      ), (
        <button
          className='cancel'
          key='cancel'
          onClick={this.onClose}
        >
          Cancel
        </button>
      )
    ];
  }

  public render() {
    const { isAddVersionDialogShowing } = this.props.appState;
    const dirOptions = { webkitdirectory: 'true' };

    return (
      <Dialog
        isShowing={isAddVersionDialogShowing}
        isShowingBackdrop={true}
        buttons={this.buttons}
        onClose={this.onClose}
        isCentered={true}
        className='add-version-dialog'
        key='add-version-dialog'
      >
        <label
          htmlFor='custom-electron-version'
          className='force-button'
        >
          Select the folder containing {getElectronNameForPlatform()}
        </label>
        <input
          type='file'
          onChange={this.onChangeFile}
          id='custom-electron-version'
          name='custom-electron-version'
          {...dirOptions}
        />
        {this.renderPath()}
      </Dialog>
    );
  }

  private renderPath(): JSX.Element | null {
    const { file, isValidElectron } = this.state;

    if (!file || !file.path) return null;

    const info = isValidElectron
      ? `We found an ${getElectronNameForPlatform()} in this folder.`
      : `We did not find a ${getElectronNameForPlatform()} in this folder...`;

    return (
      <>
        <input
          readOnly={true}
          value={file.path}
        />
        <p>
          {info}
        </p>
        {this.renderVersionInput()}
      </>
    );
  }

  private renderVersionInput(): JSX.Element | null {
    const { isValidElectron, isValidVersion, version } = this.state;

    if (!isValidElectron) return null;

    return (
      <>
        <p>
          Please specify a version, used for typings and the name.
          Must be <code>semver</code> compliant.
        </p>
        <input
          className={isValidVersion ? '' : 'hasError'}
          value={version}
          onChange={this.onChangeVersion}
          placeholder='4.0.0'
        />
      </>
    );
  }

}
