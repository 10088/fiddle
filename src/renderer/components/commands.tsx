import * as React from 'react';
import { observer } from 'mobx-react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faTerminal } from '@fortawesome/fontawesome-free-solid';

import { Runner } from './runner';
import { VersionChooser } from './version-chooser';
import { AppState } from '../state';
import { AddressBar } from './address-bar';
import { PublishButton } from './publish-button';

export interface CommandsProps {
  appState: AppState;
}

/**
 * The command bar, containing all the buttons doing
 * all the things
 *
 * @class Commands
 * @extends {React.Component<CommandsProps, {}>}
 */
@observer
export class Commands extends React.Component<CommandsProps, {}> {
  constructor(props: CommandsProps) {
    super(props);
  }

  public render() {
    const { appState } = this.props;

    return (
      <div className='commands'>
        <div>
          <Runner appState={appState} />
          <VersionChooser appState={appState} />
        </div>
        <div>
          <AddressBar appState={appState} />
          <PublishButton appState={appState} />
          <button className='button' onClick={() => appState.toggleConsole()}>
            <FontAwesomeIcon icon={faTerminal} />
          </button>
        </div>
      </div>
    );
  }
}
