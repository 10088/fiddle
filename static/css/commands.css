.commands {
  display: flex;
  justify-content: space-between;
  height: 40px;
  padding-top: 12px;
  font-family: var(--default-fonts);
}

.commands button,
.commands input,
.commands select {
  -webkit-app-region: no-drag;
}

.commands>div {
  display: flex;
}

.commands>div:first-of-type>button {
  margin-left: 80px;
}

.commands>div:last-of-type>button {
  margin-left: 5px;
}

.commands>div:last-of-type {
  margin-right: 10px;
}

.commands select {
  width: 100px;
  font-size: 12px;
  height: 28px;
  background-color: white;
  margin-left: 5px;
  line-height: 15px;
}

.commands .address-bar {
  height: 28px;
  width: 360px;
  transition: width .4s;
  display: flex;
}

.commands .address-bar.empty {
  width: 200px;
}

.commands .address-bar:focus-within span,
.commands .address-bar:focus-within input {
  border-color: #9feaf9;
}

.commands .address-bar input {
  flex-grow: 1;
  background: white;
  border-left: 0;
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
  padding-left: 0;
  height: 28px;
}

.commands .address-bar input:focus {
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
  border-left: 0;
}

.commands .address-bar span {
  background: white;
  color: #1d1d1d;
  border-bottom-color: rgb(209, 209, 209);
  border-bottom-style: solid;
  border-bottom-left-radius: 4px;
  border-bottom-width: 1px;
  border-image-outset: 0px;
  border-image-repeat: stretch;
  border-image-slice: 100%;
  border-image-source: none;
  border-image-width: 1;
  border-left-color: rgb(209, 209, 209);
  border-left-style: solid;
  border-left-width: 1px;
  border-top-color: rgb(209, 209, 209);
  border-top-left-radius: 4px;
  border-top-style: solid;
  border-top-width: 1px;
  color: rgb(0, 0, 0);
  cursor: auto;
  display: inline-block;
  font-family: system-ui;
  font-size: 11px;
  height: 28px;
  letter-spacing: normal;
  line-height: 16px;
  margin-bottom: 15px;
  margin-left: 0px;
  margin-right: 0px;
  margin-top: 0px;
  padding-bottom: 6px;
  padding-left: 10px;
  padding-right: 10px;
  padding-top: 6px;
  user-select: text;
  width: 132px;
  -webkit-app-region: no-drag;
  text-rendering: auto;
  text-shadow: none;
  text-transform: none;
}