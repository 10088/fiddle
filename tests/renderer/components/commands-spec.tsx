import { shallow } from 'enzyme';
import * as React from 'react';

import { Commands } from '../../../src/renderer/components/commands';

jest.mock('../../../src/renderer/components/commands-runner', () => ({
  Runner: 'runner'
}));

jest.mock('../../../src/renderer/components/commands-version-chooser', () => ({
  VersionChooser: 'version-chooser'
}));

jest.mock('../../../src/renderer/components/address-bar', () => ({
  AddressBar: 'address-bar'
}));

jest.mock('../../../src/renderer/components/publish-button', () => ({
  PublishButton: 'publish-button'
}));

describe('Commands component', () => {
  let store: any;

  beforeEach(() => {
    store = {
      gistId: null
    };
  });

  it('renders', () => {
    const wrapper = shallow(<Commands appState={store} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('opens the console on console click', () => {
    store.toggleConsole = jest.fn();

    const wrapper = shallow(<Commands appState={store} />);
    wrapper.find('button').simulate('click');
    expect(store.toggleConsole).toHaveBeenCalled();
  });
});
