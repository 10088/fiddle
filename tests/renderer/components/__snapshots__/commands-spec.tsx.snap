// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Commands component renders 1`] = `
<div
  className="commands"
>
  <div>
    <runner
      appState={
        Object {
          "gistId": null,
        }
      }
    />
    <version-chooser
      appState={
        Object {
          "gistId": null,
        }
      }
    />
  </div>
  <div>
    <address-bar
      appState={
        Object {
          "gistId": null,
        }
      }
    />
    <publish-button
      appState={
        Object {
          "gistId": null,
        }
      }
    />
    <button
      className="button"
      onClick={[Function]}
    >
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              640,
              512,
              Array [],
              "f120",
              "M257.981 272.971L63.638 467.314c-9.373 9.373-24.569 9.373-33.941 0L7.029 444.647c-9.357-9.357-9.375-24.522-.04-33.901L161.011 256 6.99 101.255c-9.335-9.379-9.317-24.544.04-33.901l22.667-22.667c9.373-9.373 24.569-9.373 33.941 0L257.981 239.03c9.373 9.372 9.373 24.568 0 33.941zM640 456v-32c0-13.255-10.745-24-24-24H312c-13.255 0-24 10.745-24 24v32c0 13.255 10.745 24 24 24h304c13.255 0 24-10.745 24-24z",
            ],
            "iconName": "terminal",
            "prefix": "fas",
          }
        }
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        symbol={false}
        title=""
        transform={null}
      />
    </button>
  </div>
</div>
`;
